substrate/primitives/runtime/src/traits/transaction_extension/mod.rs

```diff
@@ -0,0 +1,526 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! The transaction extension trait.
+
+use crate::{
+	scale_info::{MetaType, StaticTypeInfo},
+	transaction_validity::{TransactionValidity, TransactionValidityError, ValidTransaction},
+	DispatchResult,
+};
+use codec::{Codec, Decode, Encode};
+use impl_trait_for_tuples::impl_for_tuples;
+#[doc(hidden)]
+pub use sp_std::marker::PhantomData;
+use sp_std::{self, fmt::Debug, prelude::*};
+use sp_weights::Weight;
+use tuplex::{PopFront, PushBack};
+
+use super::{DispatchInfoOf, Dispatchable, OriginOf, PostDispatchInfoOf};
+
+mod as_transaction_extension;
+mod dispatch_transaction;
+#[allow(deprecated)]
+pub use as_transaction_extension::AsTransactionExtension;
+pub use dispatch_transaction::DispatchTransaction;
+
+/// Shortcut for the result value of the `validate` function.
+pub type ValidateResult<Val, Call> =
+	Result<(ValidTransaction, Val, OriginOf<Call>), TransactionValidityError>;
+
+/// Simple blanket implementation trait to denote the bounds of a type which can be contained within
+/// a [`TransactionExtension`].
+pub trait TransactionExtensionInterior:
+	Codec + Debug + Sync + Send + Clone + Eq + PartialEq + StaticTypeInfo
+{
+}
+impl<T: Codec + Debug + Sync + Send + Clone + Eq + PartialEq + StaticTypeInfo>
+	TransactionExtensionInterior for T
+{
+}
+
+/// Base for [TransactionExtension]s; this contains the associated types and does not require any
+/// generic parameterization.
+pub trait TransactionExtensionBase: TransactionExtensionInterior {
```

Comment:
Why do we need this? I couldn't find anything where this is really required to be separate? Asking because I think that having two traits makes this more complicated to handle, as you need to implement both traits always.